docs/design/run.md

```diff
@@ -0,0 +1,275 @@
+# Introducing JJ run
+
+Authors: [Philip Metzger](mailto:philipmetzger@bluewin.ch), [Martin von Zweigberk](mailto:martinvonz@google.com), [Danny Hooper](mailto:hooper@google.com), [Waleed Khan](mailto:me@waleedkhan.name)
+
+Initial Version, 10.12.2022 (view full history [here](https://docs.google.com/document/d/14BiAoEEy_e-BRPHYpXRFjvHMfgYVKh-pKWzzTDi-v-g/edit))
+
+
+**Summary:** This Document documents the design of a new `run` command for 
+Jujutsu which will be used to seamlessly integrate with build systems, linters
+and formatters. This is achieved by running a user-provided command or script 
+across multiple revisions. For more details, read the 
+[Use-Cases of jj run.](#Use-Cases-of-jj-run)
+
+## Preface
+
+The goal of this Design Document is to specify the correct behavior of `jj run`.
+The points we decide on here I (Philip Metzger) will try to implement. There 
+exists some prior work in other DVCS:
+* `git test`: part of git-branchless. Similar to this proposal for `jj run? . 
+* `hg run`: Google's internal Mercurial extension. Similar to this proposal for
+`jj run`.
+Details not available. 
+* `hg fix`: Google's open source Mercurial extension: [source code][fix-src]. A
+more specialized approach to rewriting file content without full context of the
+working directory. 
+* `git rebase -x`: runs commands opportunistically as part of rebase. 
+* `git bisect run`: run a command to determine which commit introduced a bug.
+
+## Context and Scope:
+
+The initial need for some kind of command runner integrated in the VCS, surfaced
+in a [github discussion][pre-commit]. In a [discussion][hooks] on discord about the 
+git-hook model, there was consensus about not repeating their mistakes.
+
+For `jj run` there is prior art in Mercurial, git branchless and Google's 
+internal Mercurial. Currently git-branchless `git test` and `hg fix` implement
+some kind of command runner. While the Google internal `hg run` works in 
+conjunction with CitC (Clients in the Cloud) which allows it to lazily apply
+the current command to any affected file. The base Jujutsu backend does not
+have a fancy virtual filesystem supporting it, so we can't apply this 
+optimization. 
+
+## Goals and Non-Goals: 
+
+### Goals:
+
+* We should be able to apply the command to any revision, published or unpublished.
+* We should be able to parallelize running the actual command, while preserving a
+good console output.
+* The run command should be able to work in the working copy. 
```

Comment:
> Do you mean that we'd have some flag like `jj run --run-in-working-copy 'cargo fmt'`?

The original intention was to support `jj run 'cargo clippy --fix'` in the working copy. So adjusting the current wc-commit and then giving it back to the users control. I do not expect it to be a different flag, as it would make wc-commit behavior special. 

> I guess that could be useful if you have a complex command set up and it's easier to run the same thing in the working copy like that than it is to type the command. Or maybe you want to debug the command by checking if it works when run in the main working copy.

Yes, thats the goal. 

> Ah, reading further down, it seems like that's not what you mean.

Where do you infer that from? This is the version we agreed upon the end of last year. 

>  How about writing this as "The command that's being run should be given a full working copy to run in"? I also avoided "The run command" there because it can very easily be misread as "The `jj run` command".

I think this is correct, but am unsure if it still encompasses the use-case above.

Reply:
I think I was just confused by the use of "the working copy", which I interpret as meaning the working copy where you normally run your editor (and build etc.). I think simply replacing "the working copy" by "a working copy" would help a lot.

Oh, or maybe you're saying that the command should be able to run on the working-copy *commit*? If that's what you meant here, I think it would be clearer as "It should be possible to run `jj run` on the working-copy commit (like it is with all other `jj` commands).". But I think it would be clearer to just remove the item in that case. I think everyone would expect that if `jj run -r @-` works, then so will `jj run -r @`.

Reply:
Rewritten to `jj run should work on a working copy`

Reply:
> Rewritten to `jj run should work on a working copy`

What do you mean by that? That the command we subprocess to will run in a temporary working copy? Or that `jj run` should work on the working-copy commit? I understand that both are true (and the latter is too obvious to mention IMO). Sorry if it feels like I'm asking the same thing over and over (because I am). I'm just trying to clarify it to other readers.

Reply:
> What do you mean by that?

Probably that `jj run` should work on any working copy, the working-copy commit itself or any other commit. 

> That the command we subprocess to will run in a temporary working copy? Or that `jj run` should work on the working-copy commit?

I mean both, as it could be implemented as either.

Reply:
We discussed my wrong terminology usage in a private chat. The new sentence should match the common terminology now.